packages/ruleset/src/functions/use-date-based-format.js

```diff
@@ -0,0 +1,359 @@
+/**
+ * Copyright 2024 IBM Corporation.
+ * SPDX-License-Identifier: Apache2.0
+ */
+
+const {
+  isArraySchema,
+  isDateSchema,
+  isDateTimeSchema,
+  isIntegerSchema,
+  isObject,
+  isObjectSchema,
+  isStringSchema,
+  schemaHasConstraint,
+  validateSubschemas,
+} = require('@ibm-cloud/openapi-ruleset-utilities');
+
+const {
+  LoggerFactory,
+  isDateBasedName,
+  isDateBasedValue,
+  isParamContentSchema,
+  isParamSchema,
+  isRequestBodySchema,
+  isResponseSchema,
+  isSchemaProperty,
+} = require('../utils');
+
+let ruleId;
+let logger;
+
+/**
+ * The implementation for this rule makes assumptions that are dependent on the
+ * presence of the following other rules:
+ *
+ * - oas3-valid-media-example
+ * - oas3-valid-schema-example
+ *
+ * These rules verify that the correct, specific format (date vs date-time) is
+ * used for schemas based on their example value. So, we aren't as specific
+ * with that check in this rule - we recommend either "date" or "date-time".
+ */
+
+module.exports = function (schema, _opts, context) {
+  if (!logger) {
+    ruleId = context.rule.name;
+    logger = LoggerFactory.getInstance().getLogger(ruleId);
+  }
+
+  return checkForDateBasedFormat(
+    schema,
+    context.path,
+    context.documentInventory.resolved
+  );
+};
+
+/**
+ * This function implements a rule that enforces date-based schemas use either
+ * the "date" or "date-time" format, so that they're accurately documented as
+ * date-based logical types. We use a heuristic based on either the name of a
+ * schema (derived from the property name, if the schema is a property schema)
+ * or the example value provided for a given schema or schema property.
+ *
+ * The logic here recursively checks all schemas for the presence of unmarked
+ * date-based schemas. As it traverses the schemas, it compiles a list of
+ * potentially-relevant example values. This way, if an object schema defines
+ * its own example, which includes a value for a nested property that should
+ * be identified by the rule, we can track down the value once we reach the
+ * schema for said property. The logic will also gather any relevant parameter
+ * or media type examples that may be defined outside of the schema path.
+ *
+ * @param {object} s the schema to check
+ * @param {array} p the array of path segments indicating the "location" of the schema within the API definition
+ * @param {object} apidef the resolved API definition
+ * @returns an array containing the violations found or [] if no violations
+ */
+function checkForDateBasedFormat(s, p, apidef) {
+  const examples = [];
+  const propertyPath = [];
+
+  return validateSubschemas(s, p, (schema, path) => {
```

Comment:
I probably should have made this clearer in the docs for `validateSubschemas()` but it _usually_ isn't what you want. What I said was this:

> This function is useful when a certain **syntactic practice** is prescribed or proscribed within **every simple schema** composed and/or nested by a schema.

Compare to the docs for `validateNestedSchemas()`:

> This function is useful when a certain **semantic quality** is required for every **composite schema** nested by a schema. The `validate` function passed to this function must itself be "composition-aware." Usually this means it will use other utilities (such as `getSchemaType()` or `schemaHasConstraint()`) to determine the semantic qualities of a composite schema.

What am I getting at here? Well, because `validateSubschemas()` treats all referenced schemas (whether they are schemas nested in properties or composed by applicators) the same, it's difficult and confusing to relate your validation logic to an outside context because the context changes between calls of the `validate` function in _irregular_ ways. We're not traversing either the logical property tree _or_ the logical composition tree. We're traversing both simultaneously!

For this reason, it really is best to limit the use of `validateSubschemas()` to validation logic whether the context _simply doesn't matter_: or, as stated in the docs, where you only care about validating a syntactic practice within a "simple schema" and you don't care about how it logically nests or composes into anything else.

This descent confusion is the cause of the bug noted in my three-part comment.

What you need to do here instead is use `validateNestedSchemas()`. You could start by calling `getPropertyNamesForSchema()` and checking the property names. Then things get a little trickier.

I would recommend factoring out the core logic from `getPropertyNamesForSchema()` into `collectFromCompositeSchema(schema, collector: (composedSchema) => collectedItems`. Then `getPropertyNamesForSchema(schema)` can be simplified into `collectFromCompositeSchema(schema, isObject(schema.properties) ? Object.keys(schema.properties) : []`. Then you can add `getExamplesForSchema(schema)` also using `collectFromCompositeSchema()`.

You can still index examples as you go so that nested schemas can refer to examples in their parents, but matching logical paths to schema paths is going to be tricky. I think I would recommend updating `validateNestedSchemas()` to track a `logicalPath` in addition to a `path`. It should be fairly straightforward, though you'll have to decide whether to break compatibility by inserting it into the middle of the current parameter list.

One subtle aspect of what I'm proposing here is that because we'd be flattening examples across logical paths, a specific example value might not apply to a particular instance of a schema — but given this is just a heuristic and a given logical path certainly ought not to have a mixture of fundamental types (\*cough\*), I think it should work fine in this case.